File: 1372. Longest ZigZag Path in a Binary Tree.py
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        self.res = 0
        def dfs(node, is_left, count):
            if node:
                self.res = max(self.res, count)
                if is_left:
                    dfs(node.left, False, 1)
                    dfs(node.right, True, count+1)
                else:
                    dfs(node.right, True, 1)
                    dfs(node.left, False, count+1)
        dfs(root, True, 0)
        dfs(root, False, 0)
        return self.res

File: 1372. Longest ZigZag Path in a Binary Tree.kt
class Solution {
    var res = 0
    fun longestZigZag(root: TreeNode?): Int {
        dfs(root, true, 0)
        dfs(root, false, 0)
        return res
    }
    
    fun dfs(node: TreeNode?, is_left: Boolean, count: Int) {
        if (node != null) {
            res = maxOf(res, count)
            if (is_left) {
                dfs(node.left, false, 1)
                dfs(node.right, true, count+1)
            } else {
                dfs(node.right, true, 1)
                dfs(node.left, false, count+1)
            }
        }
    }
}
